mlir/lib/Dialect/EmitC/IR/EmitC.cpp

```diff
@@ -929,22 +930,44 @@ LogicalResult emitc::VerbatimOp::verify() {
   return success();
 }
 
+static ParseResult parseVariadicFmtArgs(AsmParser &p,
+                                        SmallVector<Type> &params) {
+  Type type;
+  if (p.parseType(type))
+    return failure();
+
+  params.push_back(type);
+  while (succeeded(p.parseOptionalComma())) {
+    if (p.parseType(type))
+      return failure();
+    params.push_back(type);
+  }
+
+  return success();
+}
+
+static void printVariadicFmtArgs(AsmPrinter &p, ArrayRef<Type> params) {
+  llvm::interleaveComma(params, p, [&](Type type) { p.printType(type); });
+}
+
 /// Parse a format string and return a list of its parts.
 /// A part is either a StringRef that has to be printed as-is, or
 /// a Placeholder which requires printing the next operand of the VerbatimOp.
 /// In the format string, all `{}` are replaced by Placeholders, except if the
-/// `{` is escaped by `{{` - then it doesn't start a placeholder.
-FailureOr<SmallVector<emitc::VerbatimOp::ReplacementItem>>
-emitc::VerbatimOp::parseFormatString() {
-  SmallVector<ReplacementItem> items;
+/// `{` is escaped by `{{` - then it doesn't start a placeholder
+template <typename Op, class ArgType>
+FailureOr<SmallVector<typename Op::ReplacementItem>>
```

Comment:
Duplicating the ReplacementItem/Placeholder type per op is odd.

You can move them into the emitc namespace (into EmitC.h), which should be available when the table-gen generated ops are included.